docs: update function groups documentation (#1025)
NVIDIA/NeMo-Agent-Toolkit: docs/source/extend/function-groups.md

# Writing Custom Function Groups

:::{note}
Read the [Function Groups](../workflows/function-groups.md) guide first to understand how function groups work before creating your own.
:::

This guide shows you how to create custom function groups for the NVIDIA NeMo Agent toolkit. Function groups bundle related functions that share configuration, resources, and runtime context.

## When to Write a Custom Function Group

Create a custom function group when you need to:

- **Share expensive resources**: Database connections, API clients, cache instances, or connection pools across multiple related functions
- **Bundle related operations**: Group CRUD operations, file operations, or API endpoints that belong together
- **Centralize configuration**: Manage credentials, endpoints, and settings in one place for multiple functions
- **Create reusable components**: Package functionality that can be used across multiple workflows
- **Namespace functions**: Organize functions into logical groups (such as `db.query`, `db.insert`, `api.get`, `api.post`)

> ## What You'll Learn

Reviewer: This section isn't necessary. Level 2 headers should be indicated in the "On this page" list on the top right. Can remove.

    database: str = Field(description="Database name")
    user: str = Field(description="Database user")
    password: str = Field(description="Database password")
    max_connections: int = Field(default=10, description="Maximum pool size")
```

These fields become available in your YAML configuration:

```yaml
function_groups:
  db:
    _type: database_group
    host: "localhost"
    port: 5432
    database: "mydb"
    user: "${DB_USER}"
    password: "${DB_PASSWORD}"
    max_connections: 20
```

### Controlling Function Exposure

Use optional `include` or `exclude` lists to control which functions are exposed:

```python
class APIGroupConfig(FunctionGroupBaseConfig, name="api_group"):
    """Configuration for API operations."""
    base_url: str = Field(description="API base URL")
    api_key: str = Field(description="API authentication key")
    
    # Optional: specify which functions to expose
    include: list[str] = Field(
        default_factory=list,
        description="Functions to expose globally"
    )
    
    # Or alternatively, specify which to hide
    exclude: list[str] = Field(
        default_factory=list,
        description="Functions to keep private"

>     )

Author: They are but they could override the default programmatically

```

:::{note}
`include` and `exclude` are mutually exclusive. If both are provided, a `ValueError` will be raised.
:::

> **When to use each**:

Reviewer: ```suggestion
When to use `include`, `exclude`, or neither:
```

                return [dict(row) for row in rows]
        
        async def execute_fn(sql: str) -> str:
            """Execute a SQL statement (INSERT, UPDATE, DELETE)."""
            async with pool.acquire() as conn:
                await conn.execute(sql)
                return "Statement executed successfully"
        
        async def count_fn(table: str) -> int:
            """Count rows in a table."""
            async with pool.acquire() as conn:
                result = await conn.fetchval(f"SELECT COUNT(*) FROM {table}")
                return result
        
        # Add all functions to the group
        group.add_function(name="query", fn=query_fn, description=query_fn.__doc__)
        group.add_function(name="execute", fn=execute_fn, description=execute_fn.__doc__)
        group.add_function(name="count", fn=count_fn, description=count_fn.__doc__)
        
        # Yield within the context manager to keep pool alive
        yield group
    # Pool automatically closes when workflow ends
```

**Why use context managers**:
- Resource lifecycle management (automatic cleanup)
- Connection pooling efficiency
- Proper error handling
- Prevents resource leaks

> ### Using Configuration Values

Reviewer 2: This section should be earlier

    # Create authenticated HTTP client using config
    async with httpx.AsyncClient(
        base_url=config.base_url,
        headers={"Authorization": f"Bearer {config.api_key}"},
        timeout=30.0
    ) as client:
        group = FunctionGroup(config=config, instance_name="api")
        
        async def get_user_fn(user_id: int) -> dict:
            """Get user details by ID."""
            response = await client.get(f"/users/{user_id}")
            response.raise_for_status()
            return response.json()
        
        async def create_item_fn(name: str, description: str) -> dict:
            """Create a new item."""
            response = await client.post(
                "/items",
                json={"name": name, "description": description}
            )
            response.raise_for_status()
            return response.json()
        
        group.add_function(name="get_user", fn=get_user_fn, description=get_user_fn.__doc__)
        group.add_function(name="create_item", fn=create_item_fn, description=create_item_fn.__doc__)
        
        yield group
```

## Step 3: Customizing Function Schemas

> The toolkit automatically infers input and output schemas from your function type hints. You can customize these schemas for better validation and documentation.

Reviewer 2: Isnt this just the same as normal functions. Should we just link that documentation?

- Default values with validation
- Rich error messages

### Complex Input Types

Use Pydantic models for complex structured inputs:

```python
from enum import Enum
from pydantic import BaseModel, Field, HttpUrl

class RequestMethod(str, Enum):
    """Supported HTTP methods."""
    GET = "GET"
    POST = "POST"
    PUT = "PUT"
    DELETE = "DELETE"

class APIRequestInput(BaseModel):
    """Input schema for API requests."""
    endpoint: str = Field(description="API endpoint path", pattern="^/.*")
    method: RequestMethod = Field(default=RequestMethod.GET, description="HTTP method")
    headers: dict[str, str] = Field(default_factory=dict, description="Request headers")
    body: dict | None = Field(default=None, description="Request body for POST/PUT")

async def api_request_fn(endpoint: str, method: RequestMethod = RequestMethod.GET, 
                        headers: dict[str, str] | None = None, body: dict | None = None) -> dict:
    """Make an API request."""
    # Implementation goes here
    pass

group.add_function(
    name="request",
    fn=api_request_fn,
    input_schema=APIRequestInput,
    description="Make an HTTP API request"
)
```

> ## Step 4: Working with Function Groups Programmatically

Reviewer: ```suggestion
## Step 4: Work with Function Groups Programmatically
```

    group.add_function("func1", fn1)           # Always included
    group.add_function("func2", fn2)           # Always included
    group.add_function("func3_experimental",   # Conditionally included
                      fn3, 
                      filter_fn=experimental_gate)
    group.add_function("test_func4", fn4)     # Removed in production by group filter
    group.add_function("func5", fn5)          # NOT in include list, so never accessible
    
    yield group
```

**Result in production** (`environment="production"`, `enable_experimental=False`):
1. Start with: `["func1", "func2", "func3_experimental", "test_func4"]` (include list)
2. After group filter: `["func1", "func2", "func3_experimental"]` (test_func4 removed)
3. After per-function filter: `["func1", "func2"]` (func3_experimental removed)

**Result in development** (`environment="development"`, `enable_experimental=True`):
1. Start with: `["func1", "func2", "func3_experimental", "test_func4"]` (include list)
2. After group filter: `["func1", "func2", "func3_experimental", "test_func4"]` (all pass)
3. After per-function filter: `["func1", "func2", "func3_experimental", "test_func4"]` (all pass)

> ## Best Practices

Reviewer 2: A lot of this seems duplicated with the workflows section. Can we deduplicate and have one best practices section and link?

Author: Removing from here

        group.add_function("post", post)
        yield group
```

### Pattern 3: Stateful Cache

```python
@register_function_group(config_type=CacheConfig)
async def build_cache_group(config: CacheConfig, _builder: Builder):
    # Shared cache state
    cache: dict[str, tuple[Any, float]] = {}
    
    group = FunctionGroup(config=config, instance_name="cache")
    
    async def set_value(key: str, value: Any) -> str:
        """Set a cache value with TTL."""
        cache[key] = (value, time.time() + config.ttl)
        return f"Cached: {key}"
    
    async def get_value(key: str) -> Any | None:
        """Get a cache value if not expired."""
        if key in cache:
            value, expires = cache[key]
            if time.time() < expires:
                return value
            del cache[key]
        return None
    
    async def clear_cache() -> str:
        """Clear all cache entries."""
        cache.clear()
        return "Cache cleared"
    
    group.add_function("set", set_value)
    group.add_function("get", get_value)
    group.add_function("clear", clear_cache)
    yield group
```

## Troubleshooting

### Issue: Functions Not Appearing in Workflow

> **Problem**: Functions aren't available even though they're added to the group.

Reviewer: ```suggestion
**Problem**: Functions are not available even though they are added to the group.
```

**Solution**: Check your `include` list configuration:
```yaml
function_groups:
  mygroup:
    _type: my_group
    include: [func1, func2]  # Must list functions explicitly
```

Or reference the entire group:
```yaml
workflow:
  tool_names: [mygroup]  # Use group name, not individual functions
```

### Issue: Resource Leaks

**Problem**: Database connections or other resources aren't being cleaned up.

**Solution**: Always yield within the context manager:
```python
# Correct
async with create_pool() as pool:
    group = FunctionGroup(...)
    yield group  # Inside context

# Wrong
async with create_pool() as pool:
    group = FunctionGroup(...)
yield group  # Outside context - pool already closed!
```

### Issue: Filter Not Working

> **Problem**: Filter function isn't affecting which functions are available.

Reviewer: ```suggestion
**Problem**: Filter function is not affecting available functions.
```
